tools/cross_runner/src/mha.h

```diff
@@ -0,0 +1,490 @@
+        readback_output_barrirer = CD3DX12_RESOURCE_BARRIER::Transition(output_buffer_.Get(),
+            D3D12_RESOURCE_STATE_COPY_SOURCE, D3D12_RESOURCE_STATE_UNORDERED_ACCESS);
+        command_list->ResourceBarrier(1, &readback_output_barrirer);
+
+        gpu_op::Mha mha_ref(params_.type, to_dml_data_type(params_.dt), to_dml_tensor_policy(params_.layout),
+            params_.shape_a, params_.shape_b, params_.shape_c, get_shape_output(),
+             false /*params_.b_managed*/, params_.b_transposed, dml_device_, d3d12_device_, true);
+
+        // bind descriptor heap
+        auto descriptor_heap = create_descriptor_heap(d3d12_device_, mha_ref.get_total_descriptor_count());
+        ID3D12DescriptorHeap* d3d12_descriptor_heaps[] = { descriptor_heap.Get() };
+        command_list->SetDescriptorHeaps(1, d3d12_descriptor_heaps);
+
+        mha_ref.create_binding_tables(descriptor_heap->GetCPUDescriptorHandleForHeapStart(), descriptor_heap->GetGPUDescriptorHandleForHeapStart());
+        mha_ref.record_initialize(dml_cmd_recorder_, command_list);
+        close_execute_reset_wait(d3d12_device_, command_queue, command_allocator, command_list);
+
+        command_list->SetDescriptorHeaps(1, d3d12_descriptor_heaps);
+        mha_ref.record_execute(dml_cmd_recorder_, command_list, output_buffer_.Get(), input_buffer_a_.Get());
+        close_execute_reset_wait(d3d12_device_, command_queue, command_allocator, command_list);
+
+        readback_output_barrirer = CD3DX12_RESOURCE_BARRIER::Transition(output_buffer_.Get(),
+            D3D12_RESOURCE_STATE_UNORDERED_ACCESS, D3D12_RESOURCE_STATE_COPY_SOURCE);
+        command_list->ResourceBarrier(1, &readback_output_barrirer);
+        command_list->CopyResource(readback_buffer.Get(), output_buffer_.Get());
+        close_execute_reset_wait(d3d12_device_, command_queue, command_allocator, command_list);
+
+        std::vector<std::byte> ref_untyped_result(tensor_out_bytes_width);
+        readback_mapped_ptr = nullptr;
+        readback_buffer->Map(0, nullptr, reinterpret_cast<void**>(&readback_mapped_ptr));
+        std::memcpy(ref_untyped_result.data(), readback_mapped_ptr, ref_untyped_result.size());
+        readback_buffer->Unmap(0, nullptr);
+
+
+
+        if (params_.dt == DataType::eFp32)
+        {
+            return run_conformance_check<float>(data_out, ref_untyped_result, 0.05f);
+        }
+        else if (params_.dt == DataType::eFp16)
+        {
+            return run_conformance_check<Half>(data_out, ref_untyped_result, 0.05f);
+        }
+        assert(false && "Unsupported output data type!");
+        ConformanceResult ret{};
+        return ret;
+
+    }
+    protected:
+    TensorShape get_shape_output() const
+    {
+        TensorShape ret{};
+        ret.n = get_batch();
+        ret.h = get_M();
+        ret.w = get_N();
+        return ret;
+    }
+
+    std::uint32_t get_batch() const
+    {
+        return params_.shape_a.n;
+    }
+
+
+    std::uint32_t get_M() const
```

Comment:
there are 2 gemms in MHA in total, which gemm is these "M"? If its not needed please remove it

Reply:
It looks like a naming issue. Will rename get_M into get_height/ get_N into get_width.  it is still needed to get output shape based on input information. Will rename it.